### PR TITLE
Module event register with data

### PR DESCRIPTION
**Pull request by johnfxgalea**
Enables the passing of user data upon the execution of cbs of mod load and unload events.

See #2833

### CODE REVIEW

**Comment by johnfxgalea** [comment]
Test is failing because a module is not being unloaded, and calling the cb. Any recommendation on which module I should unload, or on a better way of testing this functionality please?

**Comment by johnfxgalea** [comment]
Thanks for the review. Will get these done tonight.

**Comment by johnfxgalea** [comment]
Hmm, I am not sure what caused the fail.

**Comment by derekbruening** [comment]
BTW, generally reviewers wait for replies like "Done" (or some other reply) to each of the review comments to know which ones were addressed and which outstanding or disagreed upon.

File: ext/drmgr/drmgr.c [hunk 1]
```diff
@@ -1483,7 +1511,14 @@ drmgr_modload_event(void *drcontext, const module_data_t *info,
     for (i = 0; i < iter.num; i++) {
         if (!iter.cbs.generic[i].pri.valid)
             continue;
-        (*iter.cbs.generic[i].cb.modload_cb)(drcontext, info, loaded);
+        bool is_using_user_data = iter.cbs.generic[i].is_using_user_data;
+        void *user_data = iter.cbs.generic[i].user_data;
+        if (is_using_user_data == false)
+            (*iter.cbs.generic[i].cb.modload_cb.cb_no_user_data)(drcontext, info,
+                                                                 loaded);
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

File: ext/drmgr/drmgr.c [hunk 2]
```diff
@@ -1458,6 +1464,17 @@ drmgr_register_module_load_event_ex(void (*func)
                                    (void (*)(void)) func, priority, false, NULL);
 }
 
+DR_EXPORT
+bool
+drmgr_register_module_load_event_user_data(void (*func)(void *drcontext,
+                                            const module_data_t *info,
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

File: ext/drmgr/drmgr.c [hunk 3]
```diff
@@ -1468,6 +1485,17 @@ drmgr_unregister_module_load_event(void (*func)
                                       (void (*)(void)) func);
 }
 
+DR_EXPORT
+bool
+drmgr_unregister_module_load_event_user_data(void (*func)
+                                             (void *drcontext,
+                                             const module_data_t *info,
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done. I use vim.

File: ext/drmgr/drmgr.c [hunk 4]
```diff
@@ -1507,6 +1542,18 @@ drmgr_register_module_unload_event_ex(void (*func)
                                    (void (*)(void)) func, priority, false, NULL);
 }
 
+DR_EXPORT
+bool
+drmgr_register_module_unload_event_user_data(void (*func)
+                                             (void *drcontext,
+                                             const module_data_t *info,
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

File: ext/drmgr/drmgr.h [hunk 5]
```diff
@@ -926,6 +926,19 @@ drmgr_register_module_load_event_ex(void (*func)
...
+DR_EXPORT
+/**
+ * Registers a callback function for the module load event, which
+ * behaves just like DR's module load event dr_register_module_load_event().
+ * Allows for the passing of user input \p user_data, which is available upon
+ * the execution of the callback. \return whether successful.
+ */
+bool
+drmgr_register_module_load_event_user_data(void (*func)(void *drcontext,
+                                           const module_data_t *info,
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.h]: Done.

File: suite/tests/client-interface/drmgr-test.dll.c [hunk 6]
```diff
@@ -355,6 +377,23 @@ event_thread_exit_null_user_data(void *drcontext, void *user_data)
     dr_mutex_unlock(threadlock);
 }
 
+static void
+event_mod_load(void *drcontext, const module_data_t *mod,
+                                  bool loaded, void *user_data)
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply suite/tests/client-interface/drmgr-test.dll.c]: Done.

File: suite/tests/client-interface/drmgr-test.dll.c [hunk 7]
```diff
@@ -355,6 +377,23 @@ event_thread_exit_null_user_data(void *drcontext, void *user_data)
...
+{
+    in_event_mod_load = true;
+    CHECK(user_data == (void *) mod_user_data_test, "incorrect user data for mod load");
+}
+
+
+static void
+event_mod_unload(void *drcontext, const module_data_t *mod,
+                        void *user_data)
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply suite/tests/client-interface/drmgr-test.dll.c]: Done.

File: suite/tests/client-interface/drmgr-test.dll.c [hunk 8]
```diff
@@ -89,6 +91,10 @@ static void event_thread_init_null_user_data(void *drcontext, void *user_data);
 static void event_thread_exit_null_user_data(void *drcontext, void *user_data);
 static void event_thread_context_init(void *drcontext, bool new_depth);
 static void event_thread_context_exit(void *drcontext, bool process_exit);
+static void event_mod_load(void *drcontext, const module_data_t *mod,
+                           bool loaded, void *user_data);
+static void event_mod_unload(void *drcontext, const module_data_t *mod,
+                             void *user_data);
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply suite/tests/client-interface/drmgr-test.dll.c]: Done. I loaded the same DLL as used in events.c.

File: ext/drmgr/drmgr.c [hunk 9]
```diff
@@ -1516,6 +1563,17 @@ drmgr_unregister_module_unload_event(void (*func)
                                       (void (*)(void)) func);
 }
 
+DR_EXPORT
+bool
+drmgr_unregister_module_unload_event_user_data(void (*func)
+                                               (void *drcontext,
+                                               const module_data_t *info,
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

File: ext/drmgr/drmgr.c [hunk 10]
```diff
@@ -1530,7 +1588,13 @@ drmgr_modunload_event(void *drcontext, const module_data_t *info)
...
         if (!iter.cbs.generic[i].pri.valid)
             continue;
-        (*iter.cbs.generic[i].cb.modunload_cb)(drcontext, info);
+        bool is_using_user_data = iter.cbs.generic[i].is_using_user_data;
+        void *user_data = iter.cbs.generic[i].user_data;
+        if (is_using_user_data == false)
+            (*iter.cbs.generic[i].cb.modunload_cb.cb_no_user_data)(drcontext, info);
+        else
+            (*iter.cbs.generic[i].cb.modunload_cb.cb_user_data)(drcontext, info,
+                                                                user_data);
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

File: ext/drmgr/drmgr.h [hunk 11]
```diff
@@ -926,6 +926,19 @@ drmgr_register_module_load_event_ex(void (*func)
...
+                                           bool loaded, void *user_data),
+                                           void *user_data);
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.h]: Done.

File: ext/drmgr/drmgr.h [hunk 12]
```diff
@@ -963,6 +988,18 @@ drmgr_register_module_unload_event_ex(void (*func)
...
+ * Registers a callback function for the module unload event, which
+ * behaves just like DR's module unload event dr_register_module_unload_event().
+ * Allows for the passing of user data, \p user_data, which is available upon the
+ * execution of the callback. \return whether successful.
+ */
+bool
+drmgr_register_module_unload_event_user_data(void (*func)
+                                             (void *drcontext,
+                                             const module_data_t *info,
+                                             void *user_data), void *user_data);
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr.h]: Done.

File: ext/drmgr/drmgr.c [hunk 13]
```diff
@@ -1513,12 +1512,13 @@ drmgr_modload_event(void *drcontext, const module_data_t *info,
             continue;
         bool is_using_user_data = iter.cbs.generic[i].is_using_user_data;
         void *user_data = iter.cbs.generic[i].user_data;
-        if (is_using_user_data == false)
+        if (is_using_user_data == false){
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c]: nit: space before {

Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: @derekbruening One thing that you might not be aware of. Commit 613ff1a, of this PR, resulted in a failed build. I am not sure what caused the issue, but the test seems to unreliably fail or pass. It is not a test which seems to check my PR.

Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Ill get the space in; not a prob.

Comment (derekbruening) [reply ext/drmgr/drmgr.c]: Looking: ok it was tool.drcacheoff.burst_threads which is a known flaky test (#2092) nobody has yet marked flaky or disabled or tried to fix.

Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

**Comment by derekbruening** [comment]
Thank you for the patch, and for fixing the small style isssues.

**Pull request merged**